docs/source/en/kernel_doc/overview.md

```diff
@@ -1,3 +1,171 @@
+- **MultiScaleDeformableAttention**: For vision transformers
+- **Custom Attention**: Load community kernels for specialized attention patterns
+
+### Specialized Operations
+- **Mamba Selective Scan**: Built-in optimized kernels for Mamba SSM models
+- **Causal Convolution 1D**: Efficient causal convolutions
+
+## Quick Start
+
+### Basic Usage
+
+Enable kernels when loading any model:
+
+```python
+from transformers import AutoModelForCausalLM
+
+model = AutoModelForCausalLM.from_pretrained(
+    "meta-llama/Llama-3.2-1B-Instruct",
+    use_kernels=True,
+    device_map="cuda"
+)
+
+# Model now uses optimized kernels automatically
+output = model.generate(input_ids, max_new_tokens=50)
+```
+
+### Custom Attention Kernels
+
+Use specialized attention implementations:
+
+```python
+model = AutoModelForCausalLM.from_pretrained(
+    "meta-llama/Llama-3.2-1B-Instruct",
+    attn_implementation="kernels-community/flash-attn",
+    device_map="cuda"
+)
+```
+
+## How It Works
+
+### Automatic Kernel Replacement
+
+When `use_kernels=True`, the library:
+
+1. **Scans the model**: Identifies layers that have optimized kernel implementations
+2. **Loads kernels from Hub**: Downloads and caches kernels from Hugging Face repositories
+3. **Replaces forward methods**: Swaps standard PyTorch operations with optimized kernels
+4. **Maintains compatibility**: Ensures identical outputs while improving performance
+
+### Mode-Aware Optimization
+
+Kernels automatically adapt to your training/inference workflow:
+
+```python
+model.eval()   # Uses inference-optimized kernels
+model.train()  # Switches to training-optimized kernels
+```
+
+### Kernel Sources
+
+Kernels are distributed through Hugging Face Hub repositories in the format:
+- `org/repo:layer_name` - Latest version
+- `org/repo@v0.1.0:layer_name` - Specific version
+- Supports semantic versioning constraints
+
+**Popular Kernel Repositories:**
+- [`kernels-community/flash-attn`](https://huggingface.co/kernels-community/flash-attn2) - Flash attention implementations
+- [`kernels-community/megablocks`](https://huggingface.co/kernels-community/megablocks) - MoE optimizations
+- [`kernels-community/moe`](https://huggingface.co/RedHatAI/moe) - Llama 4 MoE layers
+- [`kernels-community/liger_kernels`](https://huggingface.co/kernels-community/liger_kernels) - RMSNorm, activation functions
+
+## Requirements
+
+- **kernels** package: `pip install kernels`
+- **Recommended**: `kernels>=0.10.2` for XPU support
+
+## Advanced Features
+
+### Training and Inference Modes
+
+Kernels support different operational modes:
+
+- **Mode.INFERENCE**: Optimized for inference workloads (batch size optimization, reduced memory)
+- **Mode.TRAINING**: Optimized for training (gradient computation, mixed precision)
+- **Mode.TORCH_COMPILE**: Compatible with `torch.compile` for JIT optimization
+
+### Device-Specific Kernels
+
+Specify different kernel implementations per device:
+
+```python
+kernel_config = KernelConfig(
```

Comment:
It'd be nice to show what you do with the `kernel_config` once its defined for more context